pkg/controller/direct/bigquerydataset/dataset_controller.go

```diff
@@ -0,0 +1,334 @@
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	bigquerypb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/cloud/bigquery/v2"
+	api "google.golang.org/api/bigquery/v2"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "bigquery-controller"
+	serviceDomain = "//bigquery.googleapis.com"
+)
+
+func init() {
+	registry.RegisterModel(krm.BigQueryDatasetGVK, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &model{config: *config}, nil
+}
+
+var _ directbase.Model = &model{}
+
+type model struct {
+	config config.ControllerConfig
+}
+
+func (m *model) service(ctx context.Context) (*api.Service, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpService, err := api.NewService(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building Dataset client: %w", err)
+	}
+	return gcpService, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.BigQueryDataset{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewBigQueryDatasetRef(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get bigquery GCP client
+	gcpService, err := m.service(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:         id,
+		gcpService: gcpService,
+		desired:    obj,
+		reader:     reader,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id         *krm.BigQueryDatasetRef
+	gcpService *api.Service
+	desired    *krm.BigQueryDataset
+	actual     *api.Dataset
+	reader     client.Reader
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting BigQueryDataset", "name", a.id.External)
+
+	parent, datasetId, err := krm.ParseBigQueryDatasetExternal(a.id.External)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse bigquery dataset full name, %w", err)
+	}
+	req := &bigquerypb.GetDatasetRequest{
```

Comment:
nit: it does not look like we need to import "bigquerypb" to construct a request. We are talking to "a.gcpService.Datasets.Get" which takes a project ID and dataset ID. Same for Update() and Delete().